src/mg/PAMI/Client/Impl/ClientImpl.php

```diff
@@ -185,9 +189,7 @@ public function open()
 	    @stream_set_blocking($this->_socket, 0);
 	    $this->_currentProcessingMessage = '';
 	    //register_tick_function(array($this, 'process'));
-	    if ($this->_logger->isDebugEnabled()) {
```

Comment:
I have given this some thought. I *can* write a series of adapters specific for PAMI in order to solve this. However, the design would be kind of ugly and quite a bit complicated.

TL;DR: The costs are high and the winnings are (relatively) low. I would strongly advice *against* re-introducing this behaviour. The reasons are:

1. It directly contradicts the goals of PSR-3, which is to provide a standardised logging solution.
1. In the default case, we don't win anything over the current situation.
1. The complexity of the implementation is high.
1. The complexity of documenting this behaviour is high.
1. Finally, these CPU cycles can be won back (and then some) by upgrading to the latest PHP versions. These usually have so many performance improvements that the cost of having these calls are dwarfed by it. Since we are targeting people who run Asterisk, we have excluded almost by definition the users on shared hosting. Therefore, the remaining users will usually be able to do such an upgrade.

### Proposed implementation

Just so you know that I don't randomly shout that it's a bad idea (:wink:), here is the implementation that I think is required to re-introduce the behaviour as it was.

We will need two extra marker interfaces (don't bother with the names just yet...) and another adapter:

 * `ILogLevelEnabledChecker`
 * `ILogLevelDisabler`
 * `PAMILoggerAdapter` (implements `Psr\Log\LoggerInterface` and `Psr\Log\LoggerAwareInterface`)

We need both, because Log4Php already has the functionality to enable and disable log levels (this not needing `ILogLevelDisabler`. Also, the `Log4PhpAdapter` needs to implement `ILogLevelEnabledChecker`.

Specific implementations of PSR-3, such as Monolog, have this functionality also, but we cannot count on that as we are programming to an interface.

We will need the `PAMILogger` adapter to convert a generic PSR-3 Logger into one that has these two interfaces. This adapter must have the default to be that it enables debug logging, otherwise it the results will be very surprising to users. A user that *does* want to disable specific log levers, must wrap it in `PAMILogger` by himself, and disable these log levels specifically.

Furthermore, we will need to allow both plain PSR-3 Loggers (which will then get wrapped in `PAMILogger` and instances of `ILogLevelEnabledChecker` (which will not be wrapped).

Finally, we have to re-introduce the checks of whether these log levels are enabled or disabled.

All in all, even though I *do* understand the reasons of wanting this functionality, I don't think that it warrants the cost, both in complexity of implementation and in complexity of documentation. In the default case (when a user just injects a PSR-3 logger), it will still need to do the concatenation anyway, because we cannot know the intention of the user upfront.